### ISSUE TITLE
Discussion: Validation of user input

### ISSUE THREAD

**Issue by robinkrahl** [d-e-s-o/nitrocli#45]
Most of the user input provided via arguments or options is subject to restrictions.  These restrictions can be fixed (e. g. a hexadecimal string must always have an even number of hex digits) or variable (e. g. the current Nitrokey Pro has 3 HOTP slots, but the next version might have 5).  To which degree should we validate such input in nitrocli (that is also validated by libnitrokey or the firmware)?

Advantages of validation in nitrocli are better error messages and failing early.  Disadvantages are that we might be less compatible or might make wrong assumptions.

Currently we perform no validation.  My suggestion would be to perform basic validation on fixed restrictions (hex string), but no validation for variable restrictions (slot count) unless libnitrokey gives us the required information (e. g. a `TOTP_SLOT_COUNT` constant).

Examples for the current error messages:

Invalid hex string:
```
$ nitrocli otp set 0 test test
Could not write OTP slot: Unknown
```
Invalid slot:
```
$ nitrocli otp set 30 test test
Could not write OTP slot: InvalidSlot
```

**Comment by d-e-s-o** [comment]
Does that sound reasonable? Should we compile a list of commands that accept user input that has restrictions on it and see whether the error codes are detailed enough and, if not, engage with the Nitrokey team?

**Comment by robinkrahl** [comment]
Yes, we can try to fix that upstream.  I’ll start a wiki page for the evaluation.

**Comment by robinkrahl** [comment]
→ [libnitrokey error handling](https://github.com/d-e-s-o/nitrocli/wiki/libnitrokey-error-handling) wiki page
